File: swift/src/Ice/Blobject.swift
// Copyright (c) ZeroC, Inc.

import Foundation
import PromiseKit

/// Base protocol for dynamic dispatch servants.
public protocol Blobject {
    /// Dispatch an incoming request.
    ///
    /// - parameter inEncaps: `Data` - The encoded in-parameters for the operation.
    ///
    /// - parameter current: `Ice.Current` - The Current object to pass to the operation.
    ///
    /// - returns: `(ok: Bool, outParams: Data)` - A tuple with the following fields:
    ///
    ///   - ok: `Bool` - True if the operation completed successfully, false if
    ///     the operation raised a user exception (in this case, outParams
    ///     contains the encoded user exception). If the operation raises an
    ///     Ice run-time exception, it must throw it directly.
    ///
    ///   - outParams: `Data` - The encoded out-paramaters and return value
    ///     for the operation. The return value follows any out-parameters.
    func ice_invoke(inEncaps: Data, current: Current) throws -> (ok: Bool, outParams: Data)
}

/// Request dispatcher for Blobject servants.
public struct BlobjectDisp: Dispatcher {
    public let servant: Blobject

    public init(_ servant: Blobject) {
        self.servant = servant
    }

    public func dispatch(_ request: IncomingRequest) -> Promise<OutgoingResponse> {
        do {
            let (inEncaps, _) = try request.inputStream.readEncapsulation()
            let result = try servant.ice_invoke(inEncaps: inEncaps, current: request.current)
            return Promise.value(
                request.current.makeOutgoingResponse(ok: result.ok, encapsulation: result.outParams))
        } catch {
            return Promise(error: error)
        }
    }
}

File: python/python/Ice/Blobject.py
# Copyright (c) ZeroC, Inc.

from abc import ABC, abstractmethod
from .Object import Object

class Blobject(Object, ABC):
    """
    Special-purpose servant base class that allows a subclass to handle synchronous Ice invocations as "blobs" of
    bytes.

    This class serves as a base for creating servants that can process encoded Ice operation arguments and return
    results or exceptions in an encoded form. Subclasses must implement the `ice_invoke` method to handle the
    invocation logic.
    """

    @abstractmethod
    def ice_invoke(self, bytes, current):
        """
        Dispatches a synchronous Ice invocation.

        The operation's arguments are encoded in the `bytes` parameter. The return value must be a tuple of two
        values: the first is a boolean indicating whether the operation succeeded (True) or raised a user exception
        (False), and the second is the encoded form of the operation's results or the user exception.

        Parameters
        ----------
        bytes : bytes
            The encoded operation arguments.
        current : Ice.Current
            The current invocation context.

        Returns
        -------
        tuple of (bool, bytes)
            A tuple containing:
                - bool True if the operation succeeded, False if it raised a user exception.
                - bytes The encoded form of the operation's results or the user exception.
        """
        pass
